File: solutions/451.sort-characters-by-frequency/sort-characters-by-frequency.java
class Solution {
    public String frequencySort(String s) {
        Map<Character, Integer> map = new HashMap<>();
        for(char c : s.toCharArray())
            map.put(c, map.getOrDefault(c, 0) + 1);
        List<Map.Entry<Character, Integer>> list = new ArrayList<>();
        list.addAll(map.entrySet());
        list.sort((a, b) -> b.getValue() - a.getValue());
        StringBuilder res = new StringBuilder();
        for(Map.Entry<Character, Integer> entry : list) 
            for(int i = 0; i < entry.getValue(); i++) 
                res.append(entry.getKey());
        return res.toString();
    }
}

File: solutions/451.sort-characters-by-frequency/sort-characters-by-frequency.cpp
class Solution {
public:
    string frequencySort(string s) {
        unordered_map<char, int> freq;
        vector<string> bucket(s.size() + 1, "");
        for(char c : s) freq[c]++;
        for(auto& it: freq) {
            int n = it.second;
            char c = it.first;
            bucket[n].append(n, c);//append n 个 c
        }
        string res;
        for(int i = s.size(); i > 0; i--) 
            if(!bucket[i].empty()) 
                res.append(bucket[i]);
        return res;
    }
};